File: lib/data/sync/sync_service.dart
import 'dart:async';

import 'package:connectivity_plus/connectivity_plus.dart';

import '../local/app_database.dart';
import '../models/body_checkin.dart';
import '../remote/supabase_service.dart';

/// Pushes pending local changes and pulls remote rows for the current user.
class SyncService {
  SyncService(this._db, this._svc);
  final AppDatabase _db;
  final SupabaseService _svc;

  StreamSubscription<List<ConnectivityResult>>? _conn;
  bool _running = false;

  void start() {
    _conn ??= Connectivity().onConnectivityChanged.listen((results) {
      final online = results.any((r) => r != ConnectivityResult.none);
      if (online) {
        unawaited(syncNow());
      }
    });
  }

  void stop() {
    _conn?.cancel();
    _conn = null;
  }

  Future<void> syncNow() async {
    if (_running) return;
    if (!_svc.isConfigured) return;
    final user = _svc.currentUser;
    if (user == null) return;
    _running = true;
    try {
      await _pushPending(user.id);
      await _pullRemote(user.id);
    } catch (_) {
      // Swallow: next trigger will retry. Real app would log this.
    } finally {
      _running = false;
    }
  }

  Future<void> _pushPending(String userId) async {
    // Profile
    final profile = await _db.getProfileLocal(userId);
    if (profile != null && profile['sync_state'] == 'pending') {
      final payload = Map<String, dynamic>.from(profile)
        ..remove('sync_state')
        ..remove('updated_at');
      await _svc.client.from('profiles').upsert(payload);
      await _db.markProfileSynced(userId);
    }

    // Check-ins (deletes first, then upserts)
    final deletedIds = await _db.deletedCheckinIds();
    if (deletedIds.isNotEmpty) {
      await _svc.client
          .from('body_checkins')
          .delete()
          .inFilter('id', deletedIds);
      for (final id in deletedIds) {
        await _db.hardDeleteCheckin(id);
      }
    }

    final pending = await _db.pendingCheckins();
    if (pending.isNotEmpty) {
      await _svc.client
          .from('body_checkins')
          .upsert(pending.map((c) => c.toUpsertMap()).toList());
      for (final c in pending) {
        await _db.markCheckinSynced(c.id);
      }
    }
  }

  Future<void> _pullRemote(String userId) async {
    final rows = await _svc.client
        .from('body_checkins')
        .select()
        .eq('user_id', userId);
    for (final row in rows as List<dynamic>) {
      final c = BodyCheckin.fromMap(row as Map<String, dynamic>);
      await _db.upsertCheckinLocal(c, syncState: 'synced');
    }

    final profileRow = await _svc.client
        .from('profiles')
        .select()
        .eq('id', userId)
        .maybeSingle();
    if (profileRow != null) {
      await _db.upsertProfileLocal(profileRow, syncState: 'synced');
    }
  }
}

File: src/data/sync/sync-service.ts
import { eq, inArray } from 'drizzle-orm';

import { getDb } from '@/data/db/client';
import {
  checkinFromSupabase,
  checkinToRow,
  checkinToSupabase,
  profileFromSupabase,
  profileToRow,
  profileToSupabase,
} from '@/data/db/mappers';
import { checkins, profiles } from '@/data/db/schema';
import { getSupabase, isSupabaseConfigured } from '@/data/supabase';

let inFlight = false;

/**
 * Push pending local rows, then pull remote rows for the user. Safe to call
 * frequently; concurrent calls are coalesced.
 */
export async function syncNow(): Promise<void> {
  if (inFlight) return;
  if (!isSupabaseConfigured) return;
  const sb = getSupabase();
  const { data: userData } = await sb.auth.getUser();
  const user = userData.user;
  if (!user) return;

  inFlight = true;
  try {
    const db = await getDb();

    // ---- profile push ----
    const profileRows = await db
      .select()
      .from(profiles)
      .where(eq(profiles.id, user.id));
    const localProfile = profileRows[0];
    if (localProfile && localProfile.syncState === 'pending') {
      await sb.from('profiles').upsert(
        profileToSupabase({
          id: localProfile.id,
          displayName: localProfile.displayName,
          sex: (localProfile.sex as never) ?? null,
          birthDate: localProfile.birthDate,
          heightCm: localProfile.heightCm,
          unitSystem: (localProfile.unitSystem as never) ?? 'metric',
          goalType: (localProfile.goalType as never) ?? null,
          goalTargetWeightKg: localProfile.goalTargetWeightKg,
          goalTargetWaistCm: localProfile.goalTargetWaistCm,
          goalTargetBodyFat: localProfile.goalTargetBodyFat,
          goalDeadline: localProfile.goalDeadline,
          updatedAt: localProfile.updatedAt,
        }),
      );
      await db
        .update(profiles)
        .set({ syncState: 'synced' })
        .where(eq(profiles.id, user.id));
    }

    // ---- checkins: deletes ----
    const deletedRows = await db
      .select({ id: checkins.id })
      .from(checkins)
      .where(eq(checkins.syncState, 'deleted'));
    if (deletedRows.length > 0) {
      const ids = deletedRows.map((r) => r.id);
      await sb.from('body_checkins').delete().in('id', ids);
      await db.delete(checkins).where(inArray(checkins.id, ids));
    }

    // ---- checkins: pending upserts ----
    const pendingRows = await db
      .select()
      .from(checkins)
      .where(eq(checkins.syncState, 'pending'));
    if (pendingRows.length > 0) {
      const payload = pendingRows.map((r) =>
        checkinToSupabase({
          id: r.id,
          userId: r.userId,
          checkinDate: r.checkinDate,
          weightKg: r.weightKg,
          bodyFatPct: r.bodyFatPct,
          chestCm: r.chestCm,
          waistCm: r.waistCm,
          hipsCm: r.hipsCm,
          neckCm: r.neckCm,
          shouldersCm: r.shouldersCm,
          bicepLeftCm: r.bicepLeftCm,
          bicepRightCm: r.bicepRightCm,
          tricepLeftCm: r.tricepLeftCm,
          tricepRightCm: r.tricepRightCm,
          forearmLeftCm: r.forearmLeftCm,
          forearmRightCm: r.forearmRightCm,
          thighLeftCm: r.thighLeftCm,
          thighRightCm: r.thighRightCm,
          calfLeftCm: r.calfLeftCm,
          calfRightCm: r.calfRightCm,
          notes: r.notes,
          photoPath: r.photoPath,
          createdAt: r.createdAt,
          updatedAt: r.updatedAt,
        }),
      );
      await sb.from('body_checkins').upsert(payload);
      await db
        .update(checkins)
        .set({ syncState: 'synced' })
        .where(
          inArray(
            checkins.id,
            pendingRows.map((r) => r.id),
          ),
        );
    }

    // ---- pull remote ----
    const { data: remoteCheckins, error: rcErr } = await sb
      .from('body_checkins')
      .select('*')
      .eq('user_id', user.id);
    if (!rcErr && remoteCheckins) {
      for (const row of remoteCheckins) {
        const c = checkinFromSupabase(row as Record<string, unknown>);
        await db
          .insert(checkins)
          .values(checkinToRow(c, 'synced'))
          .onConflictDoUpdate({
            target: checkins.id,
            set: checkinToRow(c, 'synced'),
          });
      }
    }

    const { data: remoteProfile } = await sb
      .from('profiles')
      .select('*')
      .eq('id', user.id)
      .maybeSingle();
    if (remoteProfile) {
      const p = profileFromSupabase(remoteProfile as Record<string, unknown>);
      await db
        .insert(profiles)
        .values(profileToRow(p, 'synced'))
        .onConflictDoUpdate({
          target: profiles.id,
          set: profileToRow(p, 'synced'),
        });
    }
  } catch {
    // Swallow: next trigger retries.
  } finally {
    inFlight = false;
  }
}
